Persephone/MPDClient/MPDClient.swift

```diff
@@ -0,0 +1,212 @@
+
+    self.connection = connection
+    self.status = Status(status)
+
+    fetchQueue()
+
+    self.delegate?.didUpdateState(mpdClient: self, state: self.status!.state())
+    self.delegate?.didUpdateQueue(mpdClient: self, queue: self.queue)
+    self.delegate?.didUpdateQueuePos(mpdClient: self, song: self.status!.song())
+    idle()
+  }
+
+  func disconnect() {
+    noIdle()
+    commandQueue.async { [unowned self] in
+      mpd_connection_free(self.connection)
+    }
+  }
+
+  func fetchStatus() {
+    sendCommand(command: .fetchStatus)
+  }
+
+  func fetchQueue() {
+    sendCommand(command: .fetchQueue)
+  }
+
+  func playPause() {
+    queueCommand(command: .playPause)
+  }
+
+  func stop() {
+    queueCommand(command: .stop)
+  }
+
+  func prevTrack() {
+    queueCommand(command: .prevTrack)
+  }
+
+  func nextTrack() {
+    queueCommand(command: .nextTrack)
+  }
+
+  func queueCommand(command: Command) {
+    noIdle()
+    commandQueue.async { [unowned self] in
+      self.sendCommand(command: command)
+    }
+    idle()
+  }
+
+  func sendCommand(command: Command) {
+    switch command {
+
+    // Transport commands
+    case .prevTrack:
+      sendPreviousTrack()
+    case .nextTrack:
+      sendNextTrack()
+    case .stop:
+      sendStop()
+    case .playPause:
+      sendPlay()
+
+    case .fetchStatus:
+      guard let status = mpd_run_status(connection) else { break }
+      self.status = Status(status)
+
+    case .fetchQueue:
+      self.queue = []
+      mpd_send_list_queue_meta(connection)
+
+      while let mpdSong = mpd_recv_song(connection) {
+        let song = Song(mpdSong)
+        self.queue.append(song)
+      }
+    }
+  }
+
+  func sendNextTrack() {
+    if [MPD_STATE_PLAY, MPD_STATE_PAUSE].contains(status?.state()) {
+      mpd_run_next(connection)
+    }
+  }
+
+  func sendPreviousTrack() {
+    if [MPD_STATE_PLAY, MPD_STATE_PAUSE].contains(status?.state()) {
+      mpd_run_previous(connection)
+    }
+  }
+
+  func sendStop() {
+    mpd_run_stop(connection)
+  }
+
+  func sendPlay() {
+    if status?.state() == MPD_STATE_STOP {
+      mpd_run_play(connection)
+    } else {
+      mpd_run_toggle_pause(connection)
+    }
+  }
+
+  func noIdle() {
+    mpd_send_noidle(connection)
+  }
+
+  func idle() {
+    commandQueue.async { [unowned self] in
+      mpd_send_idle(self.connection)
+      let result = mpd_recv_idle(self.connection, true)
+
+      self.handleIdleResult(result)
+    }
+  }
+
+  func handleIdleResult(_ result: mpd_idle) {
+    let mpdIdle = Idle(rawValue: result.rawValue)
+
+    if mpdIdle.contains(.queue) {
+      self.fetchQueue()
+      self.delegate?.didUpdateQueue(mpdClient: self, queue: self.queue)
+    }
+    if mpdIdle.contains(.player) {
+      self.fetchStatus()
+      self.delegate?.didUpdateState(mpdClient: self, state: self.status!.state())
+      self.delegate?.didUpdateQueuePos(mpdClient: self, song: self.status!.song())
+    }
+    if !mpdIdle.isEmpty {
+      self.idle()
+    }
```

Comment:
Can the flow go through all three `if` blocks, or should this use `else if`s?

Reply:
Yes, the flow could potential go through all three blocks, and indeed it should. The idle returns when the server has a status update for the client, and that status update could include changes to both the queue and the player state. The idle needs to happen any time a status update comes in. The reason the idle is optional is that it could be interrupted by a command and not a status update.